docs/proposals/reliable-message-delivery.md

```diff
@@ -0,0 +1,276 @@
+Unstable networks between cloud and edge can result in frequent disconnection of edge nodes and 
+this can result in loss of messages sent to edge nodes which can’t be temporarily reached. 
+This proposal addresses this problem thus ensuring reliable message delivery.
+
+### Goals
+
+- Provide a reliable message delivery mechanism between cloud and edge.
+
+### Non-goals
+- To provide HA / failover mechanism for cloudhub.
+- To address secure communication.
+- To address encryption of data stored.
+
+## Proposal
+
+Currently all the messages from the controllers go via the channel queue (which uses beehive context for messaging) 
+to the cloudhub. The cloudhub then uses the configured protocol server (websocket/quic) to send the data to edge nodes. 
+The proposal is to introduce the node level sending message queues in cloudhub, and use the ACK message 
+returned from edge nodes to ensure the message delivery is in a reliable fashion.
+
+### Use Cases
+
+- If cloudcore being restarted or offline for a while, whenever the cloudcore is back online, 
+send the latest event to the edge node (if there is any update to be sent).
+- If edgenode being restarted or offline for a while, whenever the node is back online, 
+cloudcore will sent the latest event to make it up to date.
+
+## Design Details
+
+### Message Delivery Mechanisms
+
+There are three types of message delivery mechanisms:
+
+- At-Most-Once
+- Exactly-Once
+- At-Least-Once
+
+The existing implementation (without this proposal) in KubeEdge is 
+the first approach “At-Most-Once”, which is unreliable.
+
+The second approach “Exactly-Once” is very expensive and exhibits worst performance 
+although it provides guaranteed delivery with no message loss or duplication. 
+Since KubeEdge follows Kubernetes’ eventual consistency design principles, 
+it is not a problem for the edge to receive the same message repeatedly, as long as message is the latest one.
+
+In this proposal, “At-Least-Once” is the proposed mechanism.
+
+### At-Least-Once Delivery
+
+Shown below is a design using MessageQueue and ACKs to ensure that 
+the messages are delivered from the cloud to the edge.
+
+<img src="../images/reliable-message-delivery/reliablemessage-workflow.PNG">
+
+- We use K8s CRD stores the latest resourceVersion of resource that has been sent
+ successfully to edge. When cloudcore restarts or starts normally, 
+ it will check the resourceVersion to avoid sending old messages.
+ 
+- EdgeController and devicecontroller send the messages to the Cloudhub, and MessageDispatcher will send messages 
+to corresponding NodeMessageQueue according to the node name in message.
+
+- CloudHub will sequentially send data from the NodeMessageQueue to the corresponding edge node,
+ and will also store the message ID in an ACK channel. When the ACK message from the edge node received,
+ ACK channel will trigger to save the message resourceVersion to K8s as CRD, and send the next message.
```

Comment:
about ack channel, I recommend using `sync.Cond` to  implementation it.

Reply:
We have used the sync.Map now.